cmd/abapEnvironmentRunATCCheck.go

```diff
@@ -0,0 +1,354 @@
+package cmd
+
+import (
+	"bytes"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/cookiejar"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/SAP/jenkins-library/pkg/cloudfoundry"
+	"github.com/SAP/jenkins-library/pkg/command"
+	piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+)
+
+func abapEnvironmentRunATCCheck(config abapEnvironmentRunATCCheckOptions, telemetryData *telemetry.CustomData) {
+
+	var c = command.Command{}
+
+	var err error
+
+	c.Stdout(log.Entry().Writer())
+	c.Stderr(log.Entry().Writer())
+
+	client := piperhttp.Client{}
+	cookieJar, _ := cookiejar.New(nil)
+	clientOptions := piperhttp.ClientOptions{
+		CookieJar: cookieJar,
+	}
+	client.SetOptions(clientOptions)
+
+	var details connectionDetailsHTTP
+	var abapEndpoint string
+	//If Host flag is empty read ABAP endpoint from Service Key instead. Otherwise take ABAP system endpoint from config instead
+	if err == nil {
+		details, err = checkHost(config, details)
+	}
+
+	//Fetch Xcrsf-Token
+	if err == nil {
+		//HTTP config for fetching Xcsrf-Token
+		abapEndpoint = details.URL
+		details.URL += "/sap/bc/adt/api/atc/runs/00000000000000000000000000000000"
+		details.XCsrfToken = "fetch"
+
+		credentialsOptions := piperhttp.ClientOptions{
+			Username:  details.User,
+			Password:  details.Password,
+			CookieJar: cookieJar,
+		}
+
+		client.SetOptions(credentialsOptions)
+
+		details.XCsrfToken, err = fetchXcsrfToken("GET", details, nil, &client)
+	}
+
+	var filelocation []string
+	var yamlFile []byte
+	if err == nil {
+		filelocation, err = filepath.Glob(config.AtcConfig)
+	}
+	//Parse YAML ATC run configuration as body for ATC run trigger
+
+	if err == nil {
+		filename, _ := filepath.Abs(filelocation[0])
+		yamlFile, err = ioutil.ReadFile(filename)
+	}
+	var ATCConfig ATCconfig
+	if err == nil {
+		var result []byte
+		result, err = yaml.YAMLToJSON(yamlFile)
+		json.Unmarshal(result, &ATCConfig)
+	}
+
+	var packageString string
+	var softwareComponentString string
+	if err == nil {
+		packageString, softwareComponentString, err = buildATCCheckBody(ATCConfig)
+	}
+
+	//Trigger ATC run
+	var resp *http.Response
+	var bodyString = `<?xml version="1.0" encoding="UTF-8"?><atc:runparameters xmlns:atc="http://www.sap.com/adt/atc" xmlns:obj="http://www.sap.com/adt/objectset"><obj:objectSet>` + softwareComponentString + packageString + `</obj:objectSet></atc:runparameters>`
```

Comment:
To reduce the lines of code here, you could extract the report generation to a dedicated file.

Reply:
The report generation is already extracted in the function `parseATCResult`. I couldn't see any way to reduce the LOC to max 50 for CodeClimate.

Reply:
let's discuss this via phone :-)

Reply:
Ok i refactored the code. No major changes, just extracting code to more functions.